libass/ass_parse.c

```diff
@@ -194,11 +195,21 @@ inline void change_alpha(uint32_t *var, int32_t new, double pwr)
  * \param a first value
  * \param b second value
  * \return result of multiplication
- * Parameters and result are limited by 0xFF.
+ * At least one of the parameters must be less than or equal to 0xFF.
+ * The result is less than or equal to max(a, b, 0xFF).
  */
 inline uint32_t mult_alpha(uint32_t a, uint32_t b)
 {
-    return 0xFF - (0xFF - a) * (0xFF - b) / 0xFF;
+    return a - ((uint64_t) a * b + 0x7F) / 0xFF + b;
+}
+
+static inline double subtract(long long minuend, long long subtrahend)
+{
+#if ULLONG_MAX > LLONG_MAX
+    return (unsigned long long) minuend - subtrahend;
+#else
+#error unsigned long long cannot hold the difference of two long longs
```

Comment:
OK, I’ve now removed this check entirely.

It certainly could be done in `configure`, but (1) it isn’t immediately obvious to me what the best way of implementing that is, (2) we don’t currently check any of the other assumptions we make about the platform, and, while it may not mean it’s good to add another unchecked assumption, this one is rather unlikely to break and I’d like to see all assumptions get checked simultaneously, and (3) as my commit still isolates this code in a function, the assumption is at least obvious when looking at the source code.

I’ve kept the `subtract` function (rather than inline the code) to (sort of) document that it’s doing something tricky, and I think the code that calls it actually looks quite nice too: `subtract(a, b) / subtract(c, d)` vs `(double) ((unsigned long long) a - b) / ((unsigned long long) c - d)`. (Our old code even had an extra pair of parentheses for extra parsing fun.)